src/sources/datadog/agent/traces.rs

```diff
@@ -0,0 +1,205 @@
+use futures::future;
+use http::StatusCode;
+use prost::Message;
+use std::collections::BTreeMap;
+use std::sync::Arc;
+use warp::{filters::BoxedFilter, path, path::FullPath, reply::Response, Filter, Rejection, Reply};
+
+mod dd_proto {
+    include!(concat!(env!("OUT_DIR"), "/dd_trace.rs"));
+}
+
+pub(crate) fn build_warp_filter(
+    acknowledgements: bool,
+    multiple_outputs: bool,
+    out: SourceSender,
+    source: DatadogAgentSource,
+) -> BoxedFilter<(Response,)> {
+    build_trace_filter(acknowledgements, multiple_outputs, out, source)
+        .or(build_stats_filter())
+        .unify()
+        .boxed()
+}
+
+fn build_trace_filter(
+    acknowledgements: bool,
+    multiple_outputs: bool,
+    out: SourceSender,
+    source: DatadogAgentSource,
+) -> BoxedFilter<(Response,)> {
+    warp::post()
+        .and(path!("api" / "v0.2" / "traces" / ..))
+        .and(warp::path::full())
+        .and(warp::header::optional::<String>("content-encoding"))
+        .and(warp::header::optional::<String>("dd-api-key"))
+        .and(warp::header::optional::<String>(
+            "X-Datadog-Reported-Languages",
+        ))
+        .and(warp::query::<ApiKeyQueryParams>())
+        .and(warp::body::bytes())
+        .and_then(
+            move |path: FullPath,
+                  encoding_header: Option<String>,
+                  api_token: Option<String>,
+                  reported_language: Option<String>,
+                  query_params: ApiKeyQueryParams,
+                  body: Bytes| {
+                let events = source
+                    .decode(&encoding_header, body, path.as_str())
+                    .and_then(|body| {
+                        handle_dd_trace_payload(
+                            body,
+                            source.api_key_extractor.extract(
+                                path.as_str(),
+                                api_token,
+                                query_params.dd_api_key,
+                            ),
+                            reported_language.as_ref(),
+                            &source,
+                        )
+                        .map_err(|error| {
+                            ErrorMessage::new(
+                                StatusCode::UNPROCESSABLE_ENTITY,
+                                format!("Error decoding Datadog traces: {:?}", error),
+                            )
+                        })
+                    });
+                if multiple_outputs {
+                    handle_request(events, acknowledgements, out.clone(), Some(agent::LOGS))
+                } else {
+                    handle_request(events, acknowledgements, out.clone(), None)
+                }
+            },
+        )
+        .boxed()
+}
+
+fn build_stats_filter() -> BoxedFilter<(Response,)> {
+    warp::post()
+        .and(path!("api" / "v0.2" / "stats" / ..))
+        .and_then(|| {
+            warn!(message = "/api/v0.2/stats route is yet not supported.");
```

Comment:
Should we rate limit this?

Reply:
Indeed, if vector is acting as an aggregator there is no point showing that for every requests.